### PR TITLE
Feat/password requirements register

### LINKED ISSUE #317: Weak password rejection when registering
The user should not be able to register with a weak password, in order to prevent its account to be accessed without its consent.

### PR DESCRIPTION
**Pull request by b10taj**
# Password Requirements when Registering & Tests Improvements
## Description
This PR enhances the sign-up security by implementing comprehensive password requirements with real-time validation feedback. Users registering with email/password must now meet strict password criteria (8+ characters, uppercase, lowercase, digit, and special character), with visual feedback showing which requirements are met during registration.

**Related Issue:** Closes #317

---

## Changes

**Implementation:**
- Updated `SignInViewModel` to enforce 8-character minimum password (previously 6) with comprehensive validation:
  - At least one uppercase letter
  - At least one lowercase letter
  - At least one digit
  - At least one special character
- Added `PasswordRequirementsCard` composable to `SignInScreen` that displays in register mode only
- Real-time visual feedback with checkmarks/crosses showing which requirements are met as user types
- Password requirements are validated in order with specific error messages for each violation
- Maintains backward compatibility - sign-in flow unchanged, only sign-up is affected

**Tests:**
- Added 30+ comprehensive unit tests in `SignInViewModelTests.kt` covering:
  - Each individual password requirement validation
  - Edge cases (7 characters with all requirements, various special characters, etc.)
  - Validation order and error messaging
  - Password requirements at different positions
- Added 20+ instrumentation tests in `SignInScreenTests.kt` covering:
  - Password requirements card visibility (only in register mode)
  - UI state changes when toggling between sign-in/register modes
  - Field content persistence across mode switches
  - Button states and interactions
- Removed redundant `SignInTests.kt` file (292 lines) - replaced by more comprehensive `SignInScreenTests.kt`
- Fixed 3 failing tests due to ambiguous node matching with "Register" text

---

## Testing
- [x] Unit tests pass (77+ tests in SignInViewModelTest)
- [x] Instrumentation tests pass (70+ tests in SignInScreenTests)
- [x] Manual testing completed
- [x] Coverage maintained (≥80%)

**Test summary:** 
- Added 30+ unit tests covering all password validation rules and edge cases
- Added 20+ instrumentation tests for password requirements card UI behavior
- Fixed 3 failing tests by using test tags instead of ambiguous text matchers
- Removed 20 redundant tests from deprecated SignInTests.kt file
- Net addition: ~50 new comprehensive tests

---

## Screenshots/Videos
<p align="center">
  
  
  
</p>

### Sign In Mode (No Changes)
Password requirements card is hidden - maintains existing UX for sign-in.

### Register Mode (New Feature)
Password requirements card appears showing:
- Real-time validation with visual feedback
- Checkmarks (✓) for met requirements in green
- Crosses (✗) for unmet requirements in red
- Neutral dots when no password entered yet

**Requirements Displayed:**
- At least 8 characters long
- Contains at least one uppercase letter
- Contains at least one lowercase letter
- Contains at least one number
- Contains at least one special character

---

## Checklist
- [x] Sufficient documentation and minimal inline comments
- [x] All tests passing (127+ tests total)
- [x] No new warnings
- [x] If related issue exists: issue has all labels, fields, and milestone filled

---

### Files Changed
**Main Implementation:**
- `app/src/main/java/com/swent/mapin/ui/auth/SignInViewModel.kt` - Enhanced password validation logic
- `app/src/main/java/com/swent/mapin/ui/auth/SignInScreen.kt` - Added password requirements card UI

**Tests:**
- `app/src/test/java/com/swent/mapin/signinTests/SignInViewModelTests.kt` - Added comprehensive validation tests
- `app/src/androidTest/java/com/swent/mapin/signinTests/SignInScreenTests.kt` - Enhanced with UI tests for new feature
- `app/src/androidTest/java/com/swent/mapin/signinTests/SignInTests.kt` - **DELETED** (redundant)

### Breaking Changes
None - this only affects new sign-ups. Existing users can still sign in with their current passwords.

### Migration Notes
Users attempting to register with weak passwords will now see specific error messages guiding them to meet the requirements. The UI provides clear, real-time feedback making it easy for users to create compliant passwords.

### Code Quality
- Added "Assisted by AI" comments in new composables
- Maintained consistent code style with existing codebase
- Used semantic color tokens from MaterialTheme
- Proper test organization with descriptive test names
- High test coverage with edge case handling

### Tests Improvement

- The PR #332 improves the tests of this PR, as reminded/requested by @naderbahri .

---

Assisted by AI

### CODE REVIEW

File: app/src/main/java/com/swent/mapin/ui/auth/SignInViewModel.kt
Comment (naderbahri) [inline app/src/main/java/com/swent/mapin/ui/auth/SignInViewModel.kt]: **KDoc and validation block changes** :

- **Issue**: Error message strings are duplicated and hard-coded in the ViewModel. Extract them into constants or a single sealed type representing validation failures, and expose a single localized message formatter at the UI layer. This will improve testability and localization.

- **Issue**: The ViewModel now performs rich validation before Firebase call. This is good, but it creates duplicated logic now present in UI composable and ViewModel. Centralize validation to avoid divergence as rules change.

- **Suggestion**: Add unit tests specifically for the standalone validation function that verify all edge cases including Unicode characters, long passwords, and passwords with only punctuation characters, to ensure rules are robust.

File: app/src/main/java/com/swent/mapin/ui/auth/SignInScreen.kt [hunk 1]
```diff
@@ -137,6 +144,12 @@ fun SignInScreen(
                   },
                   singleLine = true)
 
+              // Show password requirements when in register mode
+              if (isRegistering) {
+                Spacer(modifier = Modifier.height(16.dp))
+                PasswordRequirementsCard(password = password)
+              }
```
[the remark this answers is not in the document]
Comment (b10taj) [reply app/src/main/java/com/swent/mapin/ui/auth/SignInScreen.kt]: Good catch! You're absolutely right about the recomposition issue.

By extracting the password validation logic into a `PasswordValidation` data class and wrapping it with `derivedStateOf`, we ensure that the validation calculations only run when the password string actually changes, not on every recomposition of the parent composable. This significantly reduces unnecessary computation, especially as users type.

The `testTag("passwordField")` addition also makes the tests more robust by providing an explicit, stable identifier instead of relying on potentially ambiguous UI text that could appear elsewhere in the screen.

File: app/src/main/java/com/swent/mapin/ui/auth/SignInScreen.kt [hunk 2]
```diff
@@ -261,3 +274,95 @@ fun SignInScreen(
...
+            isValid = hasLowercase,
+            showStatus = password.isNotEmpty())
+        PasswordRequirementItem(
+            "Contains at least one number", isValid = hasDigit, showStatus = password.isNotEmpty())
+        PasswordRequirementItem(
+            "Contains at least one special character",
+            isValid = hasSpecialChar,
+            showStatus = password.isNotEmpty())
+      }
+}
```
[the remark this answers is not in the document]
Comment (b10taj) [reply app/src/main/java/com/swent/mapin/ui/auth/SignInScreen.kt]: **Re: Hard-coded strings & colors** Absolutely right! All UI strings have been moved to `strings.xml` and the hard-coded green color is now `MaterialTheme.colorScheme.primary`, ensuring proper theming and localization support.

**Re: Fragile theme detection** Good point. I've replaced the `isSystemInDarkTheme()` conditional with `MaterialTheme.colorScheme.surfaceVariant` which adapts automatically to theme changes without manual checks.

**Re: Duplicated validation logic** Spot on! Created `PasswordValidationUtils.kt` with a shared `validatePassword()` function. Both the UI and ViewModel now use this single source of truth, eliminating the duplication and making the validation logic easily testable.

**Re: Separate file and previews** Great suggestion! Moved `PasswordRequirementsCard` to its own file with 4 composable previews covering different states (empty, partial, valid, dark theme).

File: app/src/main/java/com/swent/mapin/ui/auth/SignInViewModel.kt
Comment (b10taj) [reply app/src/main/java/com/swent/mapin/ui/auth/SignInViewModel.kt]: **Re: Duplicated hard-coded error strings** Excellent point! I've eliminated all hard-coded strings by creating a `PasswordValidationResult` sealed class with string resource IDs. The ViewModel now calls `applicationContext.getString(validationResult.messageResId)` to get localized error messages.

**Re: Duplicated validation logic** Spot on! The validation logic is now fully centralized in `PasswordValidationUtils.kt`. Both the UI (for real-time feedback) and ViewModel (for submission validation) use the exact same `validatePassword()` function. The `PasswordValidation` data class can convert to a `PasswordValidationResult` sealed class for error handling, ensuring a single source of truth.

**Re: Edge case unit tests** Great suggestion! I'll add `PasswordValidationUtilsTest.kt` with 40+ comprehensive tests in another PR, covering:

- Unicode characters (emoji, Cyrillic, Chinese, accented letters)
- Very long passwords (1000+ characters)
- Special character edge cases (spaces, tabs, newlines, mathematical symbols, punctuation-only)
- Boundary conditions (exactly 8 characters, 7 characters)
- All sealed class result types and error message mappings

**Review by naderbahri — APPROVED** [review]
Ok, from what I understood, you'll do the changes in the following PR. I'll take a look at this. For now, we can merge what you did.
